File: middleware/error_handler.py
from fastapi import Request, HTTPException, status
from fastapi.responses import JSONResponse
from fastapi.exceptions import RequestValidationError


async def validation_exception_handler(request: Request, exc: RequestValidationError):
    errors = exc.errors()
    first_err = errors[0] if errors else {}
    field = ".".join(str(loc) for loc in first_err.get("loc", []) if loc != "body")
    msg = first_err.get("msg", "Invalid input")
    detail_message = f"Field '{field}': {msg}" if field else msg

    return JSONResponse(
        status_code=status.HTTP_400_BAD_REQUEST,
        content={
            "error": {
                "code": "VALIDATION_ERROR",
                "message": detail_message,
            }
        },
    )


async def http_exception_handler(request: Request, exc: HTTPException):
    if isinstance(exc.detail, dict) and "error" in exc.detail:
        content = exc.detail
    else:
        code = "UNAUTHORIZED" if exc.status_code == 401 else (
            "NOT_FOUND" if exc.status_code == 404 else (
                "FORBIDDEN" if exc.status_code == 403 else (
                    "CONFLICT" if exc.status_code == 409 else "REQUEST_ERROR"
                )
            )
        )
        content = {
            "error": {
                "code": code,
                "message": str(exc.detail),
            }
        }

    return JSONResponse(
        status_code=exc.status_code,
        content=content,
    )


async def global_exception_handler(request: Request, exc: Exception):
    print(f"[Unhandled Exception] {request.method} {request.url.path}: {exc}")
    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content={
            "error": {
                "code": "INTERNAL_SERVER_ERROR",
                "message": "An unexpected internal server error occurred.",
            }
        },
    )

File: middleware/errorHandler.js
/**
 * Centralized Error Handling Middleware
 * Guarantees standard JSON error shape: { error: { code, message } }
 */
function errorHandler(err, req, res, next) {
  console.error(`[Error] ${req.method} ${req.originalUrl}:`, err);

  // Handle JSON parse errors from express.json()
  if (err instanceof SyntaxError && err.status === 400 && 'body' in err) {
    return res.status(400).json({
      error: {
        code: 'INVALID_JSON',
        message: 'Malformed JSON payload in request body.',
      },
    });
  }

  const statusCode = err.statusCode || (err.status && typeof err.status === 'number' ? err.status : 500);
  const errorCode = err.code || (statusCode === 500 ? 'INTERNAL_SERVER_ERROR' : 'REQUEST_ERROR');
  const message = err.message || 'An unexpected internal error occurred.';

  res.status(statusCode).json({
    error: {
      code: errorCode,
      message,
    },
  });
}

function notFoundHandler(req, res) {
  res.status(404).json({
    error: {
      code: 'NOT_FOUND',
      message: `The requested endpoint ${req.method} ${req.originalUrl} does not exist.`,
    },
  });
}

module.exports = {
  errorHandler,
  notFoundHandler,
};
